bchrpc/server.go

```diff
@@ -1113,16 +1208,421 @@ func (s *GrpcServer) GetMerkleProof(ctx context.Context, req *pb.GetMerkleProofR
+					V1TokenAmount: slpMsg.Data.(v1parser.SlpMint).Qty,
+				}
+			} else if int(query.PrevOutVout) == slpMsg.Data.(v1parser.SlpMint).MintBatonVout {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1MintBaton{
+					V1MintBaton: true,
+				}
+			} else {
+				return nil, status.Error(codes.Aborted, "vout is not a valid SLP output")
+			}
+		} else if slpMsg.TransactionType == "GENESIS" {
+			hash := query.PrevOutHash
+			for i := len(hash) - 1; len(result.TokenId) < len(hash); i-- {
+				result.TokenId = append(result.TokenId, hash[i])
+			}
+			if query.PrevOutVout == 1 {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1TokenAmount{
+					V1TokenAmount: slpMsg.Data.(v1parser.SlpGenesis).Qty,
+				}
+			} else if int(query.PrevOutVout) == slpMsg.Data.(v1parser.SlpGenesis).MintBatonVout {
+				result.ValidityResultType = &pb.GetTrustedSlpValidationResponse_ValidityResult_V1MintBaton{
+					V1MintBaton: true,
+				}
+			} else {
+				return nil, status.Error(codes.Aborted, "vout is not a valid SLP output")
+			}
+		}
+		result.SlpTxnOpreturn = entry.SlpOpReturn
+		results[i] = result
+	}
+	resp.Results = results
+	return resp, nil
+}
+
+// GetBip44HdAddress this method will return an address based on the requested HD path
+func (s *GrpcServer) GetBip44HdAddress(ctx context.Context, req *pb.GetBip44HdAddressRequest) (*pb.GetBip44HdAddressResponse, error) {
+	xpub := req.Xpub
+	if len(xpub) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "xpub is missing in request")
+	}
+
+	if xpub[:4] != "xpub" {
+		return nil, status.Error(codes.InvalidArgument, "xpub provided does not start with 'xpub'")
+	}
+
+	masterKey, err := hdkeychain.NewKeyFromString(xpub)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	var change uint32 = 0
+	if req.Change {
+		change = 1
+	}
+
+	ext, err := masterKey.Child(change)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	extK, err := ext.Child(req.AddressIndex)
+	if err != nil {
+		fmt.Println(err)
+		return nil, status.Error(codes.InvalidArgument, "invalid xpub: "+err.Error())
+	}
+
+	pubKey, _ := extK.ECPubKey()
+	addr, _ := extK.Address(s.chainParams)
+	slpAddrStr := ""
+	if s.slpIndex != nil {
+		slpAddr, _ := goslp.NewAddressPubKeyHash(addr.Hash160()[:], s.chainParams)
+		slpAddrStr = s.chainParams.SlpAddressPrefix + ":" + slpAddr.EncodeAddress()
+	}
+
+	res := &pb.GetBip44HdAddressResponse{
+		PubKey:   pubKey.SerializeCompressed(),
+		CashAddr: s.chainParams.CashAddressPrefix + ":" + addr.EncodeAddress(),
+		SlpAddr:  slpAddrStr,
+	}
+
+	return res, nil
+}
+
+func isMaybeSlpTransaction(txn *wire.MsgTx) bool {
+	if len(txn.TxOut) > 0 {
+		lokadHex, _ := hex.DecodeString("534c5000")
+		return bytes.Contains(txn.TxOut[0].PkScript, lokadHex)
```

Comment:
Do we want to check that it contains or starts with?

Reply:
We want contains because the scriptPubKey starts with OP_RETURN and one of a few kinds of data push opcodes.  Can be 0x04 or 0x4c04 etc.